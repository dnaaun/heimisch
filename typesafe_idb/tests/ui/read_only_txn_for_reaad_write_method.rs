use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize)]
pub struct InstallationId(u64);

#[derive(Serialize, Deserialize)]
pub struct RepositoryId(u64);

#[derive(Serialize, Deserialize)]
pub struct UserId(u64);

#[derive(macros::TypesafeIdb, Serialize, Deserialize)]
pub struct Repository {
    #[idb(id)]
    id: RepositoryId,

    name: String,

    #[idb(index)]
    installation_id: InstallationId,
}

#[derive(macros::TypesafeIdb, Serialize, Deserialize)]
pub struct User {
    #[idb(id)]
    id: UserId,

    name: String,

    #[idb(index)]
    repository_id: RepositoryId,
}

fn main() {
    #[allow(unused_must_use)]
    async {
        let db = typesafe_idb::TypesafeDb::builder("just test".into())
            .with_store::<Repository>()
            .build()
            .await
            .unwrap();

        let txn = db.txn().with_store::<Repository>().ro();
        let object_store = txn.object_store::<Repository>().unwrap();

        // Should raise type error
        object_store.delete(&RepositoryId(4)).await.unwrap();
    };
}
