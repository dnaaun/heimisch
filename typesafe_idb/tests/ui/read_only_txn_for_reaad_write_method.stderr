error[E0599]: the method `delete` exists for struct `ObjectStore<'_, Repository, ReadOnly>`, but its trait bounds were not satisfied
  --> tests/ui/read_only_txn_for_reaad_write_method.rs:47:22
   |
47 |         object_store.delete(&RepositoryId(4)).await.unwrap();
   |                      ^^^^^^ method cannot be called on `ObjectStore<'_, Repository, ReadOnly>` due to unsatisfied trait bounds
   |
  ::: src/txn.rs
   |
   | pub struct ReadOnly {}
   | ------------------- doesn't satisfy `<_ as TxnMode>::SupportsReadWrite = Present`
   |
   = note: the following trait bounds were not satisfied:
           `<typesafe_idb::ReadOnly as TxnMode>::SupportsReadWrite = Present`
